src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -1226,19 +1005,73 @@ public int hashCode() {
         }
     }
 
-    private void logNodeMessage(FlowNode node) {
-        WorkflowConsoleLogger wfLogger = new WorkflowConsoleLogger(getListener());
-        String prefix = getBranchName(node);
-        if (prefix != null) {
-            wfLogger.log(String.format("[%s] %s", prefix, node.getDisplayFunctionName()));
-        } else {
-            wfLogger.log(node.getDisplayFunctionName());
+    /**
+     * Prints nodes as they appear (including block start and end nodes).
+     */
+    private final class NodePrintListener implements GraphListener.Synchronous {
+        @Override public void onNewHead(FlowNode node) {
+            NewNodeConsoleNote.print(node, getListener());
+        }
+    }
+
+    @SuppressWarnings("rawtypes")
+    @Override public AnnotatedLargeText getLogText() {
+        return LogStorage.of(asFlowExecutionOwner()).overallLog(this, !isLogUpdated());
+    }
+
+    // TODO log-related overrides pending JEP-207:
+
+    @Override public InputStream getLogInputStream() throws IOException {
+        // Inefficient but probably rarely used anyway.
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        getLogText().writeRawLogTo(0, baos);
+        return new ByteArrayInputStream(baos.toByteArray());
+    }
+
+    @Override public Reader getLogReader() throws IOException {
+        return getLogText().readAll();
+    }
+
+    @SuppressWarnings("deprecation")
+    @Override public String getLog() throws IOException {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        getLogText().writeRawLogTo(0, baos);
+        return baos.toString("UTF-8");
+    }
+
+    @Override public List<String> getLog(int maxLines) throws IOException {
+        int lineCount = 0;
+        List<String> logLines = new LinkedList<>();
+        if (maxLines == 0) {
+            return logLines;
+        }
+        try (BufferedReader reader = new BufferedReader(getLogReader())) {
+            for (String line = reader.readLine(); line != null; line = reader.readLine()) {
+                logLines.add(line);
+                ++lineCount;
+                if (lineCount > maxLines) {
+                    logLines.remove(0);
```

Comment:
🐛 And `break`!  Don't keep adding-and-removing. 

Better yet, use `line != null && lineCount <= maxLines` as part of the `for` loop conditions

Reply:
Actually... better expressed as:

```
while((line = reader.readLine()) != null && ++lineCount <= maxLines) {
    logLines.add(line)
}
```

Reply:
Ditto, all copied from core, slated to be removed in JEP-207 as the comment above says.

Reply:
If it's wrong in core, it can be fixed in core too.

Reply:
As noted elsewhere, the newer implementation in core is different, and unusable here.

Reply:
The actual explanation for the somewhat crazy logic here was done by Devin: https://github.com/jenkinsci/workflow-job-plugin/pull/27/files#r221330578 -- makes sense in light of that.